### PR TITLE
Lab 03 - Letter frequency

### PR DESCRIPTION
**Pull request by jecofang01**
Fixes #63

#### Changes proposed in this PR:

* Implement the letter frequency
* Add test cases
* Fix the comments

### CODE REVIEW

File: 03_letters/jecofang/letter_frequency_test.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	r.Equalf(expected, actual, "Unexpected output in main()")
+}
+
+var testData = []struct {
+	in      string
+	letters map[rune]int
+	sorted  []string
+}{
+	{"", map[rune]int{}, []string{}},
+	{"aba", map[rune]int{('a'): 2, ('b'): 1}, []string{"a:2", "b:1"}},
```
Comment (anzdaddy) [inline 03_letters/jecofang/letter_frequency_test.go]: `('a')` is overkill. Just use `'a'`.

File: 03_letters/jecofang/letter_frequency.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+func sortLetters(m map[rune]int) []string {
+	keys := []int{}
+	for k := range m {
+		keys = append(keys, int(k))
+	}
+	sort.Ints(keys)
+
+	s := make([]string, len(m))
+	for i, k := range keys {
+		s[i] = fmt.Sprintf("%c:%d", k, m[rune(k)])
```
Comment (anzdaddy) [inline 03_letters/jecofang/letter_frequency.go]: It's clearer to write `s = append(s, …)` (and `s := make([]string, 0, len(m))` to initialise it).

File: 03_letters/jecofang/letter_frequency.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	m := map[rune]int{}
+	for _, v := range s {
+		m[v]++
+	}
+
+	return m
+}
+
+func sortLetters(m map[rune]int) []string {
+	keys := []int{}
```
Comment (anzdaddy) [inline 03_letters/jecofang/letter_frequency.go]:
```go
keys := make([]int{}, 0, len(m))
```

File: 03_letters/jecofang/letter_frequency_test.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	r := require.New(t)
+	var buf bytes.Buffer
+	out = &buf
+
+	// When
+	main()
+
+	// Then
+	expected := strconv.Quote(`a:2
+b:1
```
Comment (anzdaddy) [inline 03_letters/jecofang/letter_frequency_test.go]: We debated this one back and forth. It's a line call, but on balance, it's probably better to write `"a:2\nb:1"`.

File: 03_letters/jecofang/letter_frequency_test.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+`)
+	actual := strconv.Quote(buf.String())
```
Comment (anzdaddy) [inline 03_letters/jecofang/letter_frequency_test.go]: Why `strconv.Quote` both strings? Could you just compare both directly?

File: 03_letters/jecofang/letter_frequency_test.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+		map[rune]int{('a'): 2, ('b'): 2, (' '): 2, ('A'): 2, ('2'): 1, ('1'): 1, ('你'): 1},
+		[]string{" :2", "1:1", "2:1", "A:2", "a:2", "b:2", "你:1"}},
+	{"ab,aAbA21你，",
+		map[rune]int{('a'): 2, ('b'): 2, (','): 1, ('A'): 2, ('2'): 1, ('1'): 1, ('你'): 1, ('，'): 1},
+		[]string{",:1", "1:1", "2:1", "A:2", "a:2", "b:2", "你:1", "，:1"}},
+}
+
+func TestLetter(t *testing.T) {
+	// Given
+	r := require.New(t)
```
Comment (anzdaddy) [inline 03_letters/jecofang/letter_frequency_test.go]: `// Given` doesn't apply here, since `r` is just part of the testing infrastructure, not a precondition for the tests. Consider that removing it and rewriting `r.Equal(…)` as `require.Equal(t, …)` would leave you with exactly the same test.

File: 03_letters/jecofang/letter_frequency_test.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+		l := letter(d.in)
+
+		// Then
+		r.Equal(d.letters, l, "unexpected output")
+	}
+}
+
+func TestSortLetters(t *testing.T) {
+	// Given
+	r := require.New(t)
```
Comment (anzdaddy) [inline 03_letters/jecofang/letter_frequency_test.go]: `// Given` doesn't apply here.

File: 03_letters/jecofang/letter_frequency_test.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+	for _, d := range testData {
+
```
Comment (anzdaddy) [inline 03_letters/jecofang/letter_frequency_test.go]: Unnecessary blank line

File: 03_letters/jecofang/letter_frequency_test.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+		map[rune]int{('a'): 2, ('b'): 2, (','): 1, ('A'): 2, ('2'): 1, ('1'): 1, ('你'): 1, ('，'): 1},
+		[]string{",:1", "1:1", "2:1", "A:2", "a:2", "b:2", "你:1", "，:1"}},
+}
+
+func TestLetter(t *testing.T) {
+	// Given
+	r := require.New(t)
+
+	for _, d := range testData {
+
```
Comment (anzdaddy) [inline 03_letters/jecofang/letter_frequency_test.go]: Unnecessary blank line

**Review by anzdaddy — DISMISSED** [review]

**Comment by juliaogris** [comment]
Can you please update your PR description to (I've done it for you for lab1 & 2):
```
Fixes #63

#### Changes proposed in this PR:

-  Implement the letter frequency
-  Add test cases
```

The issues listed as `Fixes #<ISSUE_NR>` gets automatically closed when the PR is merged.

File: 06_puppy/madhavansu/main.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
+package main
```
Comment (juliaogris) [inline 06_puppy/madhavansu/main.go]: This file shouldn't be part of the PR...

**Review by juliaogris — DISMISSED** [review]
Again, looks like a very solid solution. Please remove all but letter files from this PR (and maybe rebase -i them away from history).

**Review by camscale — APPROVED** [review]
That's looking great now. Thanks.
